### ISSUE TITLE
explicitNulls flag not working for deserialisation of objects inside lists

### ISSUE THREAD

**Issue by Harrypulvirenti** [Kotlin/kotlinx.serialization#1800]
Hi everyone!

As described in the title I found an issue with the flag `explicitNulls = false`.

The flag is working fine for deserialising object like:

```
@Serializable
data class TestDTO(
   val title: String,
   val subtitle: String?
)
```

In this case the flag is working fine meaning that, if you set it to false, there is no need to specify the default null value in the `subtitle` field.
So when you try to deserialise the `TestDTO` class and the `subtitle` field is missing in the Json response this will be deserialised as a null value instead of throwing an exception.

If you have to deserialise something like this:

```
@Serializable
data class ListDTO(
   val testList: List<TestDTO>
)

@Serializable
data class TestDTO(
   val title: String,
   val subtitle: String?
)
```

In this case trying to deserialise the `ListDTO` object with the `subtitle` missing in the JSON response the deserialiser will throw a an exception like this:

`IOErrorException(t=java.util.NoSuchElementException: Key subtitle is missing in the map.)`

**To Reproduce**
- Set the flag `explicitNulls = false`
- Have a wrapper DTO that is containing a list of DTO objects and this second DTO should have one of more fields nullable without default value specified
- Try to deserialise the wrapper of the list with a json response in which the nullable values are missing from the response

**Expected behaviour**

The Deserialiser should not throw the `IOErrorException` exception and deserialise the response with the nullable fields converted as null.

**Environment**
 - Kotlin version: 1.5.31
 - Library version: 1.3.1
 - Kotlin platforms: JVM
 - Gradle version: 7.2

**Comment by sandwwraith** [comment]
This test is working fine for me:

```
        val j = Json { explicitNulls = false }
        val foo = ListDTO(listOf(TestDTO("foo", null), TestDTO("bar", null)))
        val s = j.encodeToString(foo)
        println(s)
        val decoded = j.decodeFromString<ListDTO>(s)
        println(decoded)
```

Also, `kotlinx.serialization` does not have such thing as `IOErrorException`. Maybe you mixed some things up?

**Comment by Harrypulvirenti** [comment]
@sandwwraith After digging a bit more on this issue ( I tried to run also your example ) I found that in my case the problem is with the `Enums`.
Let me add some more details that I found:

- The real exception is `NoSuchElementException` and not `IOErrorException` which is only a generic wrapper.
- In the configuration of the Json instance, I have also this flag `coerceInputValues = true`.
- In my tests, the deserialiser was throwing the exception only removing the `= null` from the enum field. The issue was not occurring with other fields of the DTO received inside the list.
- I was not able to reproduce the issue in a test environment but only while running the project on an Android phone ( not sure if this is relevant )
